### PR TITLE
feat: split nodebuilder generics into separate states

### LINKED ISSUE #7811: Replace NodeBuilder State Generic with individual types
### Describe the feature

Currently the `NodeBuilder` use a generic to advance its state

https://github.com/paradigmxyz/reth/blob/3224837523dddadf86f4d2f27869e5b0b4fda789/crates/node-builder/src/builder.rs#L295-L295

this results in bloated type bounds.

configuring the node is effectively a one-way street
the nested TypesState generics can be replaced with a few helper structs that represent the state of the node and perhaps some helper state traits to reduce bloat

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by mattsse**
closes #7811

This adds dedicated types for the builder states, the top level API remains unchanged.

This unblocks a few more followups:
* #7812
* #7388
* custom launcher for dev mode, optimism etc...

### CODE REVIEW

File: crates/node-builder/src/builder/mod.rs [hunk 1]
```diff
@@ -0,0 +1,666 @@
...
+            .field("config", &self.config)
+            .finish()
+    }
+}
+//
+// /// The initial state of the node builder process.
+// #[derive(Debug, Default)]
+// #[non_exhaustive]
+// pub struct InitState;
+//
```
Comment (onbjerg) [inline crates/node-builder/src/builder/mod.rs:625]: do we rm this?

File: crates/node-builder/src/components/builder.rs [hunk 2]
```diff
@@ -170,31 +169,31 @@ impl Default for ComponentsBuilder<(), (), (), ()> {
...
-    ) -> impl std::future::Future<Output = eyre::Result<NodeComponents<Node, Self::Pool>>> + Send
-    {
-        self(ctx)
-    }
+    ) -> impl Future<Output = eyre::Result<Self::Components>> + Send;
 }
+
+// impl<Node, F, Fut, Pool> NodeComponentsBuilder<Node> for F
+// where
+//     Node: FullNodeTypes,
```
Comment (onbjerg) [inline crates/node-builder/src/components/builder.rs:186]: do we rm this

File: crates/node-builder/src/launch.rs [hunk 3]
```diff
@@ -0,0 +1,559 @@
...
+
+        // extract the jwt secret from the args if possible
+        let default_jwt_path = data_dir.jwt_path();
+        let jwt_secret = config.rpc.auth_jwt_secret(default_jwt_path)?;
+
+        // adjust rpc port numbers based on instance number
+        config.adjust_instance_ports();
+
+        // Start RPC servers
+
```
Comment (onbjerg) [inline crates/node-builder/src/launch.rs:514]:
```suggestion
        // Start RPC servers
```

File: crates/node-builder/src/components/builder.rs (hunk 2 above)
Comment (mattsse) [reply crates/node-builder/src/components/builder.rs]: added back

File: crates/node-builder/src/builder/mod.rs (hunk 1 above)
Comment (mattsse) [reply crates/node-builder/src/builder/mod.rs]: yep removed

**Review by onbjerg — APPROVED** [review]

**Review by DaniPopes — APPROVED** [review]

**Pull request merged**